File: _2023/src/Day1.java
import java.util.AbstractMap;
import java.util.List;
import java.util.Map;

public class Day1 extends Day {
    public Day1(String[] lines) {
        super(lines);
    }

    public int part1() {
        int calibrationValuesSum = 0;

        for (String line : lines) {
            int i = 0;
            while (!Character.isDigit(line.charAt(i))) {
                i++;
            }
            int first = line.charAt(i) - '0';

            i = line.length() - 1;
            while (!Character.isDigit(line.charAt(i))) {
                i--;
            }
            int last = line.charAt(i) - '0';

            int calibrationValue = first * 10 + last;
            calibrationValuesSum += calibrationValue;
        }

        return calibrationValuesSum;
    }

    static final Map<String, Integer> mapping = Map.ofEntries(
            new AbstractMap.SimpleEntry<>("1", 1),
            new AbstractMap.SimpleEntry<>("2", 2),
            new AbstractMap.SimpleEntry<>("3", 3),
            new AbstractMap.SimpleEntry<>("4", 4),
            new AbstractMap.SimpleEntry<>("5", 5),
            new AbstractMap.SimpleEntry<>("6", 6),
            new AbstractMap.SimpleEntry<>("7", 7),
            new AbstractMap.SimpleEntry<>("8", 8),
            new AbstractMap.SimpleEntry<>("9", 9),
            new AbstractMap.SimpleEntry<>("one", 1),
            new AbstractMap.SimpleEntry<>("two", 2),
            new AbstractMap.SimpleEntry<>("three", 3),
            new AbstractMap.SimpleEntry<>("four", 4),
            new AbstractMap.SimpleEntry<>("five", 5),
            new AbstractMap.SimpleEntry<>("six", 6),
            new AbstractMap.SimpleEntry<>("seven", 7),
            new AbstractMap.SimpleEntry<>("eight", 8),
            new AbstractMap.SimpleEntry<>("nine", 9)
    );
    static final List<String> numbers = mapping.keySet().stream().toList();

    public int part2() {
        int calibrationValuesSum = 0;

        for (String line : lines) {
            int first = -1;
            int i = 0;

            while (first == -1) {
                for (String number : numbers) {
                    if (line.startsWith(number, i)) {
                        first = mapping.get(number);
                        break;
                    }
                }
                i++;
            }

            int last = -1;
            while (last == -1) {
                for (String number : numbers) {
                    if (line.endsWith(number)) {
                        last = mapping.get(number);
                        break;
                    }
                }
                line = line.substring(0, line.length() - 1);
            }

            int calibrationValue = first * 10 + last;
            calibrationValuesSum += calibrationValue;
        }

        return calibrationValuesSum;
    }
}

File: 2023/cpp/src/day1/Day1.cpp
#include <string>
#include <cctype>

#include "Day1.h"

Day1::Day1() : Day(1, false) {
};

Day1::~Day1() = default;

int Day1::part1() {
    int sum = 0;

    for (std::string line: m_lines) {
        int val = 0;

        const int lineLength = static_cast<int>(line.length());
        for (int i = 0; i < lineLength; i++) {
            if (const char c = line[i]; std::isdigit(c)) {
                val = val * 10 + c - '0';
                break;
            }
        }

        for (int i = lineLength - 1; i >= 0; i--) {
            if (const char c = line[i]; std::isdigit(c)) {
                val = val * 10 + c - '0';
                break;
            }
        }

        sum += val;
    }

    return sum;
}

int Day1::part2() {
    int sum = 0;
    const std::string nums[] = {"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};

    for (const std::string& line: m_lines) {
        const int numsLength = 10;
        int val = 0;

        const int lineLength = static_cast<int>(line.length());

        bool found = false;
        for (int i = 0; i < lineLength && !found; i++) {
            char c = line[i];
            if (std::isdigit(c)) {
                val = val * 10 + c - '0';
                break;
            }
            if (i >= 1) {
                for (int j = 0; j < numsLength; j++) {
                    if (line.substr(0, i + 1).find(nums[j]) != std::string::npos) {
                        val = val * 10 + j;
                        found = true;
                        break;
                    }
                }
            }
        }

        found = false;
        for (int i = lineLength - 1; i >= 0 && !found; i--) {
            if (const char c = line[i]; std::isdigit(c)) {
                val = val * 10 + c - '0';
                break;
            }
            if (i >= 1) {
                for (int j = 0; j < numsLength; j++) {
                    if (line.substr(i, lineLength).find(nums[j]) != std::string::npos) {
                        val = val * 10 + j;
                        found = true;
                        break;
                    }
                }
            }
        }

        sum += val;
    }

    return sum;
}
